test/circt-reduce/memory-stubber.mlir

```diff
@@ -0,0 +1,180 @@
+      %7 = firrtl.subfield %memory_w1(0) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.uint<4>
+      %8 = firrtl.subfield %memory_w1(1) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.uint<1>
+      %9 = firrtl.subfield %memory_w1(2) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.clock
+      %10 = firrtl.subfield %memory_w0(3) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.bundle<a: uint<8>, b: sint<8>>
+      %11 = firrtl.subfield %memory_w0(4) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.bundle<a: uint<1>, b: uint<1>>
+      %12 = firrtl.subfield %memory_w0(0) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.uint<4>
+      %13 = firrtl.subfield %memory_w0(1) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.uint<1>
+      %14 = firrtl.subfield %memory_w0(2) : (!firrtl.bundle<addr: uint<4>, en: uint<1>, clk: clock, data: bundle<a: uint<8>, b: sint<8>>, mask: bundle<a: uint<1>, b: uint<1>>>) -> !firrtl.clock
+      firrtl.connect %14, %clock : !firrtl.clock, !firrtl.clock
+      firrtl.connect %13, %wEn : !firrtl.uint<1>, !firrtl.uint<1>
+      firrtl.connect %12, %wAddr : !firrtl.uint<4>, !firrtl.uint<4>
+      firrtl.connect %11, %wMask : !firrtl.bundle<a: uint<1>, b: uint<1>>, !firrtl.bundle<a: uint<1>, b: uint<1>>
+      firrtl.connect %10, %wData : !firrtl.bundle<a: uint<8>, b: sint<8>>, !firrtl.bundle<a: uint<8>, b: sint<8>>
+      firrtl.connect %9, %clock : !firrtl.clock, !firrtl.clock
+      firrtl.connect %8, %wEn : !firrtl.uint<1>, !firrtl.uint<1>
+      firrtl.connect %7, %wAddr : !firrtl.uint<4>, !firrtl.uint<4>
+      firrtl.connect %6, %wMask : !firrtl.bundle<a: uint<1>, b: uint<1>>, !firrtl.bundle<a: uint<1>, b: uint<1>>
+      firrtl.connect %5, %wData : !firrtl.bundle<a: uint<8>, b: sint<8>>, !firrtl.bundle<a: uint<8>, b: sint<8>>
+      firrtl.connect %4, %clock : !firrtl.clock, !firrtl.clock
+      firrtl.connect %3, %wAddr : !firrtl.uint<4>, !firrtl.uint<4>
+      firrtl.connect %out0, %2 : !firrtl.bundle<a: uint<8>, b: sint<8>>, !firrtl.bundle<a: uint<8>, b: sint<8>>
+      firrtl.connect %1, %clock : !firrtl.clock, !firrtl.clock
+      firrtl.connect %0, %wAddr : !firrtl.uint<4>, !firrtl.uint<4>
+      firrtl.connect %out1, %2 : !firrtl.bundle<a: uint<8>, b: sint<8>>, !firrtl.bundle<a: uint<8>, b: sint<8>>
+    }
```

Comment:
This test and the check are _MASSIVE_. Is there any way that this can be simplified to be more maintainable going forward?

E.g., I think you can test this with a 1r0w memory that has an `sint<8>` data type.  The check can then be simplified.

Reply:
Yeah, I will simplify it.  I just copied in the massive test case I was using to manually test this so we could look at it.  I think we need a read memory with a bundle type to exercise the the [recursive connect](https://github.com/llvm/circt/pull/2383/files#diff-7c1db57bf47a9aeb958fd78a01efddd9fd9181837803e6ef73792e38dabd6ae8R329) though.

Reply:
Yeah I'm also concerned about this being a pain to maintain.  It is effectively an integration test of a TON of different subsystems, and small changes to them will continually force this to be updated.

Reply:
I guess the idea is that I'm testing only a specific reduction pattern, `MemoryStubber`, and it's the only reduction allowed to run as specified by the command at the top of the file.  The code I'm trying to exercise is relatively short and simple, it should only be replacing memories with xor chains and is pretty self-contained. I simplified the tests, but I had to be very imprecise to get it that short.

If this still seems bad, let me know.  Maybe we just forgo testing?